builder/hspkg-builder.nix

```diff
@@ -39,86 +39,19 @@ let
 
-  # This is the `Cabal` library that was built for `cabal-install` to use.
-  # It makes sense to use this version (when possible) because it will match the behavior of
-  # building with `cabal-install` (including fixes that may not be in the
-  # version of Cabal bundled with GHC).
-  cabalFromCabalInstall = buildPackages.haskell-nix.cabal-install-unchecked.${compiler-nix-name}.project.hsPkgs.Cabal.components.library;
-
-  # Check there is no chance we are building `cabalFromCabalInstall`.  Using `cabalFromCabalInstall`
-  # to build itseld would cause infinite recursion.
-  useCabalFromCabalInstall =
-        # `cabalFromCabalInstall` is not cross compiled
-        stdenv.buildPlatform != stdenv.hostPlatform
-      ||
-        # These are the dependencies of `Cabal`
-        !builtins.elem package.identifier.name
-          ["nix-tools" "alex" "happy" "hscolour" "Cabal" "bytestring" "aeson" "time"
-           "filepath" "base-compat-batteries" "base-compat" "unix" "directory" "transformers"
-           "containers" "binary" "mtl" "text" "process" "parsec"];
-
-  defaultSetup = setup-builder ({
-    name = "${ghc.targetPrefix}default-Setup";
-    component = {
-      depends = config.setup-depends ++ lib.optional useCabalFromCabalInstall cabalFromCabalInstall;
-      libs = [];
-      frameworks = [];
-      doExactConfig = false;
-      includeDirs = [];
-      asmSources = [];
-      cSources = [];
-      cmmSources = [];
-      cxxSources = [];
-      jsSources = [];
-      extraSrcFiles = [ "Setup.hs" "Setup.lhs" ];
-      pkgconfig = [];
-      build-tools = [];
-
-      platforms = null;
-      preBuild = null;   postBuild = null;
-      preInstall = null; postInstall = null;
-      preUnpack = null;  postUnpack = null;
-    };
-    package = {
-      identifier = {
-        name = "default-Setup";
-        version = "1.0";
-      };
-      homepage = null;
-      synopsis = null;
-      license = "MIT";
-    };
-    src = null;
-    cleanSrc = buildPackages.runCommand "default-Setup-src" {} ''
-      mkdir $out
-      cat ${defaultSetupSrc} > $out/Setup.hs
-    '';
-    inherit defaultSetupSrc;
-  } // lib.optionalAttrs useCabalFromCabalInstall {
-    # This is needed so that we don't get duplicate packages when we
-    # add a custom Cabal package to the dependencies.  That way custom
-    # setups won't complain about e.g. binary from the Cabal dependencies
-    # and binary from the global package-db.
-    nonReinstallablePkgs = [];
-  });
-
-  # buildPackages.runCommand "default-Setup" { nativeBuildInputs = [(ghc.passthru.buildGHC or ghc)]; } ''
-  #   cat ${defaultSetupSrc} > Setup.hs
-  #   mkdir -p $out/bin
-  #   ${(ghc.passthru.buildGHC or ghc).targetPrefix}ghc Setup.hs --make -o $out/bin/Setup
-  # '';
-
   setup = if package.buildType == "Simple"
-    then defaultSetup
-    else setup-builder {
+    then ghc.defaultSetupFor package.identifier.name
```

Comment:
It would allow the project to select a newer or older version of `Cabal` the library if we put this on cabal-install (since haskell.nix could pick the default setup from the cabal-install passed to the project).

Reply:
Is that good or bad? I think probably good?

Reply:
Yes good.  I'll try to set this up.  I think we might need to keep the version of default setup built with Cabal from GHC on the ghc derivation (probably makes sense to have it there and for booting I think it might be necessary).